### PR TITLE
composite roles

### PR DESCRIPTION
**Pull request by haraldsegliens**
I found out that this tool couldn't configure composite roles. Therefore added the possibility to configure composite roles.

Will also update documentation when  [https://github.com/mayope/keycloakmigration/pull/31](url) pull request will be completed because there could be merge issues about docs folder.

### CODE REVIEW

File: src/main/kotlin/de/klg71/keycloakmigration/changeControl/actions/role/AddRoleAction.kt [hunk 1]
```diff
@@ -38,8 +44,25 @@ class AddRoleAction(
         }
         // If the action fails from here we have to role it back
         setExecuted()
-        findRole().run {
-            client.updateRole(updateRole(this), id, realm())
+        val role = findRole()
+        client.updateRole(updateRole(role), role.id, realm())
+        if(composite != null && composite && !compositeChildRoles.isNullOrEmpty()) {
```
Comment (klg71) [inline src/main/kotlin/de/klg71/keycloakmigration/changeControl/actions/role/AddRoleAction.kt:49]: this statement can be shortened to `composite == true`

File: src/main/kotlin/de/klg71/keycloakmigration/changeControl/actions/role/AddRoleAction.kt [hunk 2]
```diff
@@ -38,8 +44,25 @@ class AddRoleAction(
...
+            val roleItems = compositeChildRoles.map {
+                val foundRole = if (it.clientId == null) {
```
Comment (klg71) [inline src/main/kotlin/de/klg71/keycloakmigration/changeControl/actions/role/AddRoleAction.kt:51]: i would like to refactor the `compositeChilldRoles.map{}`-block into a separate method

File: src/test/kotlin/de/klg71/keycloakmigration/changeControl/actions/role/AddRoleIntegTest.kt [hunk 3]
```diff
@@ -32,4 +33,25 @@ class AddRoleIntegTest : AbstractIntegrationTest() {
                 .hasMessage("Role with name: integrationTest already exists in realm: ${testRealm}!")
 
     }
+
+    @Test
+    fun testAddCompositeRoles() {
```
Comment (klg71) [inline src/test/kotlin/de/klg71/keycloakmigration/changeControl/actions/role/AddRoleIntegTest.kt:38]: :+1:

File: src/main/kotlin/de/klg71/keycloakmigration/changeControl/actions/role/AddRoleAction.kt (hunk 1 above)
Comment (haraldsegliens) [reply src/main/kotlin/de/klg71/keycloakmigration/changeControl/actions/role/AddRoleAction.kt]: done

File: src/main/kotlin/de/klg71/keycloakmigration/changeControl/actions/role/AddRoleAction.kt (hunk 2 above)
Comment (haraldsegliens) [reply src/main/kotlin/de/klg71/keycloakmigration/changeControl/actions/role/AddRoleAction.kt]: done. check whether this is better

**Review by klg71 — APPROVED** [review]

**Comment by klg71** [comment]
Merged and will it as soon as possible if its released :)

**Comment by klg71** [comment]
Released with  https://github.com/mayope/keycloakmigration/releases/tag/0.2.22

**Pull request merged**